pkg/manager/member/tikv_scaler.go

```diff
@@ -52,91 +53,187 @@ func (s *tikvScaler) Scale(meta metav1.Object, oldSet *apps.StatefulSet, newSet
 }
 
 func (s *tikvScaler) ScaleOut(meta metav1.Object, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
-	_, ordinal, replicas, deleteSlots := scaleOne(oldSet, newSet)
-	resetReplicas(newSet, oldSet)
-	obj, ok := meta.(runtime.Object)
+	tc, ok := meta.(*v1alpha1.TidbCluster)
 	if !ok {
-		return fmt.Errorf("cluster[%s/%s] can't conver to runtime.Object", meta.GetNamespace(), meta.GetName())
+		klog.Errorf("tikvScaler.ScaleOut: failed to convert cluster %s/%s, scale out will do nothing", meta.GetNamespace(), meta.GetName())
+		return nil
 	}
-	klog.Infof("scaling out tikv statefulset %s/%s, ordinal: %d (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinal, replicas, deleteSlots.List())
-	var pvcName string
-	switch meta.(type) {
-	case *v1alpha1.TidbCluster:
-		pvcName = fmt.Sprintf("tikv-%s-tikv-%d", meta.GetName(), ordinal)
-	default:
-		return fmt.Errorf("tikv.ScaleOut, failed to convert cluster %s/%s", meta.GetNamespace(), meta.GetName())
+
+	scaleOutParallelism := 1
+	if tc.Spec.TiKV.ScaleOutParallelism != nil {
+		scaleOutParallelism = int(*tc.Spec.TiKV.ScaleOutParallelism)
 	}
-	_, err := s.deps.PVCLister.PersistentVolumeClaims(meta.GetNamespace()).Get(pvcName)
-	if err == nil {
-		_, err = s.deleteDeferDeletingPVC(obj, v1alpha1.TiKVMemberType, ordinal)
+	_, ordinals, replicas, deleteSlots := scaleMulti(oldSet, newSet, scaleOutParallelism)
+	klog.Infof("scaling out tikv statefulset %s/%s, ordinal: %v (replicas: %d, delete slots: %v)", oldSet.Namespace, oldSet.Name, ordinals, replicas, deleteSlots.List())
+
+	scaleOutOne := func(ordinal int32) error {
```

Comment:
IMHO, this function is too long, do we need to split it as a method like `tikvScaler.scaleOutOne()` but not closure?

same advice for TiFlash.

Reply:
Thanks for the review, I will check it later, when I finish the e2e test.

Reply:
I have refactored the code. 
PTAL.